app/app.go

```diff
@@ -164,41 +173,98 @@ func (app *MerkleEyesApp) CheckTx(tx []byte) abci.Result {
 	return app.doTx(tree, tx)
 }
 
+// nonce prefix
+func nonceKey(nonce []byte) []byte {
+	return append([]byte("/nonce/"), nonce...)
+}
+
+// key prefix
+func storeKey(key []byte) []byte {
+	return append([]byte("/key/"), key...)
+}
+
+// function by which we interact with database
 func (app *MerkleEyesApp) doTx(tree merkle.Tree, tx []byte) abci.Result {
 	if len(tx) == 0 {
 		return abci.ErrEncodingError.SetLog("Tx length cannot be zero")
 	}
-	typeByte := tx[0]
-	tx = tx[1:]
-	switch typeByte {
-	case WriteSet: // Set
-		key, n, err := wire.GetByteSlice(tx)
-		if err != nil {
-			return abci.ErrEncodingError.SetLog(cmn.Fmt("Error reading key: %v", err.Error()))
+
+	fmt.Println(string(tx))
+
+	txStruct := new(TxStruct)
+	err := json.Unmarshal(tx, &txStruct)
+	if err != nil {
+		return abci.ErrEncodingError.SetLog(cmn.Fmt("Error decoding JSON: %v", err))
+	}
+
+	key := []byte(txStruct.Key)
+	value := []byte(txStruct.Value)
+	compareValue := []byte(txStruct.CompareValue)
+	nonce := []byte(txStruct.Nonce)
+
+	// every tx requires a nonce
+	if len(nonce) == 0 {
+		return abci.ErrEncodingError.SetLog(cmn.Fmt("Nonce cannot be blank"))
+	}
+
+	// search for nonce
+	_, _, exists := tree.Get(nonceKey(nonce))
+	if exists {
+		return abci.ErrBadNonce.AppendLog(fmt.Sprintf("Nonce %s already exists", string(nonce)))
+	}
+
+	// set nonce
+	tree.Set(nonceKey(nonce), []byte("found"))
+
+	// responses for all request methods
+	switch txStruct.Method {
+	case "get":
+		if len(key) == 0 {
+			return abci.ErrEncodingError.SetLog(cmn.Fmt("Key cannot be blank"))
+		}
+		_, value, exists := tree.Get(storeKey(key))
+		if exists {
+			fmt.Println("GET", cmn.Fmt("%s", string(key)), cmn.Fmt("%s", string(value)))
+			return abci.OK.SetData(value)
+		} else {
+			return abci.ErrBaseUnknownAddress.AppendLog(fmt.Sprintf("Cannot find key: %s", string(key)))
 		}
-		tx = tx[n:]
-		value, n, err := wire.GetByteSlice(tx)
-		if err != nil {
-			return abci.ErrEncodingError.SetLog(cmn.Fmt("Error reading value: %v", err.Error()))
+
+	case "set":
+		if len(key) == 0 {
+			return abci.ErrEncodingError.SetLog(cmn.Fmt("Key cannot be blank"))
 		}
-		tx = tx[n:]
-		if len(tx) != 0 {
-			return abci.ErrEncodingError.SetLog(cmn.Fmt("Got bytes left over"))
+		if len(value) == 0 {
+			return abci.ErrEncodingError.SetLog(cmn.Fmt("Value cannot be blank"))
 		}
+		tree.Set(storeKey(key), value)
+
+	case "remove":
+		// TODO: implement "key not found"
```

Comment:
no key validation? maybe we move key validation outside the switch statement since its common to all